src/netlink.c

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2019 Xaptum, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "netlink.h"
+
+#include <net/if.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+int
+enftun_netlink_read_message(struct enftun_netlink* nl, char* buf, size_t buflen)
+{
+    nl->io_vector.iov_base = buf;
+    nl->io_vector.iov_len = buflen;
+
+    ssize_t bytes_in_msg = recvmsg(nl->fd, &nl->msg, MSG_DONTWAIT);
+    if (bytes_in_msg < 0 || (nl->msg.msg_namelen != sizeof(nl->sock_addr)))
+        return -1;
+
+    nl->io_vector.iov_base = NULL;
+    nl->io_vector.iov_len = 0;
+
+    return 0;
```

Comment:
Return the number of bytes read (`bytes_in_message`).  Otherwise the caller won't know how much of `buf` was populated.